## Разработка приложения для прогнозирования соотношения "матрица-наполнитель"

Создадим два варианта приложения:
* 1. веб-приложение
* 2. консольное

In [2]:
# Импортируем необходимые библиотеки
import numpy as np
import tensorflow as tf
from tensorflow import keras

import flask
from flask import Flask, request, render_template

from tensorflow.keras.models import load_model
from joblib import load

#### Веб-приложение Flask

In [3]:
# Инициализируем приложение Flask
app = Flask(__name__)

In [4]:
# Загружаем модель и масштабаторы
nn_model = keras.models.load_model('C:/Users/Asus/Documents/УЧЕБА_Data Science/0. ВКР/Application//model_matrix/')
scaler_x = load('/contence/app/minmax_scl_x.pkl')
scaler_y = load('/contence/app/minmax_scl_y.pkl')

In [5]:
# Определяем маршрут приложения для страницы веб-приложения по умолчанию
@app.route('/')
def home():
    return render_template('main.html')

In [6]:
# Создаем новый маршрут приложения, который считывает ввод из формы «main.html» 
# и при нажатии кнопки "Рассчитать" выводит результат
@app.route('/predict', methods = ['POST'])
def predict():
    int_features = [float(x) for x in request.form.values()]
    X = scaler_x.transform(np.array(int_features).reshape(1,-1))
    prediction = nn_model.predict(X)
    output = scaler_y.inverse_transform(prediction) 
    return render_template('main.html', 
                           prediction_text = 'Прогнозное значение соотношения "матрица - наполнитель": {}'.format(output[0][0]))

In [7]:
# Запуск сервера Flask
if __name__ == "__main__":
    app.run(debug=True, use_reloader=False)

 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   Use a production WSGI server instead.
 * Debug mode: on


 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
127.0.0.1 - - [25/Apr/2023 18:04:22] "GET / HTTP/1.1" 200 -


1/1 [==============================] - 2s 2s/step


127.0.0.1 - - [25/Apr/2023 18:04:22] "POST /predict HTTP/1.1" 200 -


Приложение выдает значение 2.914088010787964

Соотношение "матрица-наполнитель" для аналогичных показателей (позиция 3 датафрейма df_clean) составляет 2.767918

Введенные параметры:
* Соотношение матрица-наполнитель 2.767918
* Плотность, кг/м3 2000.000000
* Модуль упругости, ГПа 748.000000
* Количество отвердителя, м.% 111.860000
* Содержание эпоксидных групп,%_2 22.267857
* Температура вспышки, С_2 284.615385
* Поверхностная плотность, г/м2 210.000000
* Модуль упругости при растяжении, ГПа 70.000000
* Прочность при растяжении, МПа 3000.000000
* Потребление смолы, г/м2 220.000000
* Угол нашивки 0.000000
* Шаг нашивки 5.000000
* Плотность нашивки 60.000000

#### Консольное приложение

In [8]:
# Создадим функцию для ввода данных
def input_variable():
    x1 = float(input('Плотность, кг/м3: '))
    x2 = float(input('Модуль упругости, ГПа: '))
    x3 = float(input('Количество отвердителя, м.%: '))
    x4 = float(input('Содержание эпоксидных групп,%_2: '))
    x5 = float(input('Температура вспышки, С_2: '))
    x6 = float(input('Поверхностная плотность, г/м2: '))
    x7 = float(input('Модуль упругости при растяжении, ГПа: '))
    x8 = float(input('Прочность при растяжении, МПа: '))
    x9 = float(input('Потребление смолы, г/м2: '))
    x10 = float(input('Угол нашивки: '))
    x11 = float(input('Шаг нашивки: '))
    x12 = float(input('Плотность нашивки: '))
    return x1, x2, x3, x4, x5, x6, x7, x8, x9, x10, x11, x12

In [9]:
# Создадим функцию для вызова приложения
def app_model():
    # Загружаем модель и масштабаторы
    nn_model = load_model('model_matrix')
    scaler_x = load('minmax_scl_x.pkl')
    scaler_y = load('minmax_scl_y.pkl')
   
    print('Приложение прогнозирует соотношение "матрица-наполнитель"')
    for i in range(110):
        try:
            print('Введите "1" для прогноза, "2" для выхода')
            check = input()
      
            if check == '1':
                print('Введите данные для прогноза')
                X = input_variable()
                X =  scaler_x.transform(np.array(X).reshape(1,-1))
                prediction = nn_model.predict(X)
                output = scaler_y.inverse_transform(prediction) 
                print('Прогнозное значение соотношения "матрица-наполнитель": ')
                print(output[0][0])

            elif check == '2':
                break
            else:
                print('Повторите выбор')
                
        except Exception as e:
            print(e)
            print('Введены некорректные данные. Пожалуйста, повторите операцию')
            
# Запускаем приложение
app_model()

Приложение прогнозирует соотношение "матрица-наполнитель"
Введите "1" для прогноза, "2" для выхода
1
Введите данные для прогноза
Плотность, кг/м3: 2000
Модуль упругости, ГПа: 748
Количество отвердителя, м.%: 111.860000
Содержание эпоксидных групп,%_2: 22.267857
Температура вспышки, С_2: 284.615385
Поверхностная плотность, г/м2: 210
Модуль упругости при растяжении, ГПа: 70
Прочность при растяжении, МПа: 3000
Потребление смолы, г/м2: 220
Угол нашивки: 0
Шаг нашивки: 5
Плотность нашивки: 60
1/1 [==============================] - 0s 62ms/step
Прогнозное значение соотношения "матрица-наполнитель": 
2.914088
Введите "1" для прогноза, "2" для выхода
2


Приложение выдает значение 2.914088

Соотношение "матрица-наполнитель" для аналогичных показателей (позиция 3 датафрейма df_clean) составляет 2.767918

Введенные параметры:
* Соотношение матрица-наполнитель 2.767918
* Плотность, кг/м3 2000.000000
* Модуль упругости, ГПа 748.000000
* Количество отвердителя, м.% 111.860000
* Содержание эпоксидных групп,%_2 22.267857
* Температура вспышки, С_2 284.615385
* Поверхностная плотность, г/м2 210.000000
* Модуль упругости при растяжении, ГПа 70.000000
* Прочность при растяжении, МПа 3000.000000
* Потребление смолы, г/м2 220.000000
* Угол нашивки 0.000000
* Шаг нашивки 5.000000
* Плотность нашивки 60.000000